File: Data Structures & Algorithms/search-2d-matrix/submission-11.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        left = 0
        right = len(matrix) - 1
        i = -1

        while left <= right:
            mid = int((left + right) / 2)

            if target == matrix[mid][0]:
                return True
            elif matrix[mid][0] < target:
                print(mid)
                if mid != len(matrix) - 1:
                    if matrix[mid + 1][0] > target:
                        i = mid
                        break
                    else:
                        left = mid + 1
                else:
                    i = mid
                    break
            elif matrix[mid][0] > target:
                right = mid - 1
        
        if i < 0:
            return False
        
        arr = matrix[i]
        l = 0
        r = len(arr) - 1

        while l <= r:
            mid = int((l + r) / 2)
                
            if arr[mid] < target:
                l = mid + 1
            elif arr[mid] > target:
                r = mid - 1
            else:
                return True
        
        return False

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-11.java
class Solution {
    public int search(int[] nums, int target) {
        return bisectionSearch(0, nums.length - 1, nums, target);
    }

    public int bisectionSearch(int left, int right, int[] nums, int target) {
        int mid = left + (right - left) / 2;

        if (left > right) {
            return -1;
        }

        if (nums[mid] == target) {
            return mid;
        }

        if (nums[left] <= nums[mid]) {
            // it on the left
            if (nums[left] <= target && target <= nums[mid]) {
                return bisectionSearch(left, mid, nums, target);
            } else {
                return bisectionSearch(mid + 1, right, nums, target);
            }
        } else {
            if (nums[mid] <= target && target <= nums[right]) {
                return bisectionSearch(mid + 1, right, nums, target);
            } else {
                return bisectionSearch(left, mid, nums, target);
            }
        }
    }
}
